File: ProcedualArt/serialization/serialize.ts
import { Matrix, Polygon, Polyline, Vector } from "@/Core/geometry";
import { Json } from "@/Core/utils";
import {
    Graph,
    internal_is_length_graph,
    internal_is_vertex_graph
} from "@/ProcArt/graph";
import { Grid, InternalGrid } from "@/ProcArt/grid";
import { GrayImage, RGBImage } from "@/ProcArt/image";
import { Sketch } from "Core/sketch/sketch";
import {
    serialize_matrix,
    serialize_polygon,
    serialize_polyline,
    serialize_vector
} from "./geometry";
import {
    serialize_length_graph,
    serialize_shape_graph,
    serialize_vertex_graph
} from "./graph";
import {
    serialize_boolean_grid,
    serialize_matrix_grid,
    serialize_number_grid,
    serialize_vec3_grid,
    serialize_vector_grid
} from "./grid";
import { serialize_gray_image, serialize_rgb_image } from "./image";
import { stringify_f64_array, stringify_u8_array } from "./number_array";
import { serialize_sketch } from "./sketch";
import { StoffSerializable, StoffSerializableTag } from "./types";

export function serialize(r: StoffSerializable, max_depth = 5): string {
    return JSON.stringify(serialize_to_json(r, max_depth));
}

export function serialize_to_json(
    r: StoffSerializable,
    max_depth = 5
): {
    type: StoffSerializableTag;
    data: Json;
} {
    if (max_depth == 0) {
        throw new Error("Nested to deep!");
    }

    if (r instanceof Float64Array) {
        return {
            type: "number_array",
            data: stringify_f64_array(r)
        };
    }

    if (r instanceof Uint8Array) {
        return {
            type: "number_array",
            data: stringify_u8_array(r)
        };
    }

    if (typeof r == "string") {
        return {
            type: "string",
            data: r
        };
    }

    if (typeof r == "number") {
        return {
            type: "number",
            data: r
        };
    }

    if (typeof r == "boolean") {
        return {
            type: "boolean",
            data: r
        };
    }

    if (r == null) {
        return {
            type: "null",
            data: r
        };
    }

    if (r instanceof RGBImage) {
        return serialize_rgb_image(r);
    }

    if (r instanceof GrayImage) {
        return serialize_gray_image(r);
    }

    if (r instanceof Polyline) {
        return serialize_polyline(r);
    }

    if (r instanceof Polygon) {
        return serialize_polygon(r);
    }

    if (r instanceof Vector) {
        return serialize_vector(r);
    }

    if (r instanceof Matrix) {
        return serialize_matrix(r);
    }

    if (r instanceof Sketch) {
        return serialize_sketch(r);
    }

    if (r instanceof Grid) {
        const s = r as typeof r & InternalGrid;
        if (s.type == "number") {
            return serialize_number_grid(s);
        }

        if (s.type == "vector") {
            return serialize_vector_grid(s);
        }

        if (s.type == "vec3") {
            return serialize_vec3_grid(s);
        }

        if (s.type == "boolean") {
            return serialize_boolean_grid(s);
        }

        return serialize_matrix_grid(s);
    }

    if (r instanceof Graph) {
        if (internal_is_length_graph(r)) {
            return serialize_length_graph(r);
        }

        if (internal_is_vertex_graph(r)) {
            return serialize_vertex_graph(r);
        }

        return serialize_shape_graph(r);
    }

    if (Array.isArray(r)) {
        if (r.every((a) => typeof a == "number")) {
            return {
                type: "number_array",
                data: stringify_f64_array(r as number[])
            };
        }

        if (r.every((a) => a instanceof Vector)) {
            return {
                type: "vector_array",
                data: stringify_f64_array(
                    (r as Vector[]).flatMap((v) => [v.x, v.y])
                )
            };
        }

        return {
            type: "array",
            data: r.map((a) => serialize_to_json(a, max_depth - 1))
        };
    }

    return {
        type: "object",
        data: Object.fromEntries(
            Object.entries(r).map(([key, value]) => [
                key,
                serialize_to_json(value, max_depth - 1)
            ])
        )
    };
}

File: Python/stoff_types/src/stoff_types/serialization/serialize.py
import json
from dataclasses import asdict, is_dataclass
from typing import Any, TypeAlias, cast

import numpy as np

from ..types.graph import Graph
from ..types.grid import Grid
from .geometry import (
    serialize_matrix,
    serialize_polygon,
    serialize_polyline,
    serialize_vector,
)
from .graph import serialize_length_graph, serialize_shape_graph, serialize_vertex_graph
from .grid import (
    serialize_boolean_grid,
    serialize_matrix_grid,
    serialize_number_grid,
    serialize_vec3_grid,
    serialize_vector_grid,
)
from .image import serialize_gray_image, serialize_rgb_image
from .index import (
    BooleanGrid,
    GrayImage,
    LengthGraph,
    Matrix,
    MatrixGrid,
    NumberGrid,
    Polygon,
    Polyline,
    RGBImage,
    ShapeGraph,
    Sketch,
    StoffSerializable,
    Vec3Grid,
    Vector,
    VectorGrid,
    VertexGraph,
)
from .number_array import stringify_f64_array, stringify_u8_array
from .sketch import serialize_sketch

StoffSerializableTag: TypeAlias = str


def serialize(r: StoffSerializable, max_depth: int = 5) -> str:
    return json.dumps(serialize_to_json(r, max_depth))


def serialize_to_json(
    r: StoffSerializable,
    max_depth: int = 5,
) -> Any:
    if max_depth == 0:
        raise ValueError("Nested too deep!")

    if isinstance(r, np.ndarray) and r.dtype == np.uint8:
        return {
            "type": "uint8_array",
            "data": stringify_u8_array(r),
        }

    if isinstance(r, np.ndarray):
        if not np.issubdtype(r.dtype, np.number):
            raise TypeError(
                "Only numeric NumPy arrays can be serialized as number_array"
            )

        return {
            "type": "number_array",
            "data": stringify_f64_array(r),
        }

    if isinstance(r, str):
        return {
            "type": "string",
            "data": r,
        }

    if isinstance(r, bool):
        return {
            "type": "boolean",
            "data": r,
        }

    if isinstance(r, (int, float)):
        return {
            "type": "number",
            "data": r,
        }

    if r is None:
        return {
            "type": "null",
            "data": None,
        }

    if isinstance(r, RGBImage):
        return serialize_rgb_image(r)

    if isinstance(r, GrayImage):
        return serialize_gray_image(r)

    if isinstance(r, Polyline):
        return serialize_polyline(r)

    if isinstance(r, Polygon):
        return serialize_polygon(r)

    if isinstance(r, Vector):
        return serialize_vector(r)

    if isinstance(r, Matrix):
        return serialize_matrix(r)

    if isinstance(r, Sketch):
        return serialize_sketch(r)

    if isinstance(r, Grid):
        if r.type == "number":
            return serialize_number_grid(cast("NumberGrid", r))

        if r.type == "vector":
            return serialize_vector_grid(cast("VectorGrid", r))

        if r.type == "vec3":
            return serialize_vec3_grid(cast("Vec3Grid", r))

        if r.type == "boolean":
            return serialize_boolean_grid(cast("BooleanGrid", r))

        return serialize_matrix_grid(cast("MatrixGrid", r))

    if isinstance(r, Graph):
        if r.type == "vertex_graph":
            return serialize_vertex_graph(cast("VertexGraph", r))

        if r.type == "length_graph":
            return serialize_length_graph(cast("LengthGraph", r))

        return serialize_shape_graph(cast("ShapeGraph", r))

    if isinstance(r, list):
        if all(isinstance(value, (int, float)) for value in r):
            return {
                "type": "number_array",
                "data": stringify_f64_array(r),
            }

        if all(isinstance(value, Vector) for value in r):
            values: list[float] = []

            for vector in r:
                values.extend((vector.x, vector.y))

            return {
                "type": "vector_array",
                "data": stringify_f64_array(values),
            }

        return {
            "type": "array",
            "data": [serialize_to_json(value, max_depth - 1) for value in r],
        }

    if isinstance(r, tuple):
        return {
            "type": "array",
            "data": [serialize_to_json(value, max_depth - 1) for value in r],
        }

    if isinstance(r, dict):
        return {
            "type": "object",
            "data": {
                str(key): serialize_to_json(value, max_depth - 1)
                for key, value in r.items()
            },
        }

    if is_dataclass(r):
        return {
            "type": "object",
            "data": {
                key: serialize_to_json(value, max_depth - 1)
                for key, value in asdict(r).items()
            },
        }

    if hasattr(r, "__dict__"):
        return {
            "type": "object",
            "data": {
                key: serialize_to_json(value, max_depth - 1)
                for key, value in vars(r).items()
            },
        }

    raise TypeError(f"Unsupported serializable type: {type(r).__name__}")
